vast/vast.cpp

```diff
@@ -11,472 +11,69 @@
-                        || !sys_cfg.openssl_key.empty()
-                        || !sys_cfg.openssl_passphrase.empty()
-                        || !sys_cfg.openssl_capath.empty()
-                        || !sys_cfg.openssl_cafile.empty();
-  VAST_INFO(banner() << '\n');
-  scoped_actor self{sys};
-  actor node;
-  if (spawn_node) {
-    VAST_INFO("spawning local node:", id);
-    node = self->spawn(system::node, id, abs_dir);
-    if (start.opts.count("bare") == 0) {
-      // If we're not in bare mode, we spawn all core actors.
-      auto spawn_component = [&](auto&&... xs) {
-        return [&] {
-          auto result = error{};
-          auto args = make_message(std::move(xs)...);
-          self->request(node, infinite, "spawn", std::move(args)).receive(
-            [](const actor&) { /* nop */ },
-            [&](error& e) { result = std::move(e); }
-          );
-          return result;
-        };
-      };
-      auto err = error::eval(
-        spawn_component("metastore"),
-        spawn_component("archive"),
-        spawn_component("index"),
-        spawn_component("importer")
-      );
-      if (err) {
-        VAST_ERROR(self->system().render(err));
-        self->send_exit(node, exit_reason::user_shutdown);
-        return 1;
-      }
-    }
-  } else {
-    auto host = node_endpoint.host;
-    if (node_endpoint.host.empty())
-      node_endpoint.host = "127.0.0.1";
-    VAST_INFO("connecting to", node_endpoint.host << ':' << node_endpoint.port);
-    auto remote_actor = [&]() -> expected<actor> {
-      if (use_encryption)
-#ifdef VAST_USE_OPENSSL
-        return openssl::remote_actor(sys, node_endpoint.host,
-                                     node_endpoint.port);
-#else
-        return make_error(ec::unspecified, "not compiled with OpenSSL support");
-#endif
-      auto& mm = sys.middleman();
-      return mm.remote_actor(node_endpoint.host, node_endpoint.port);
-    };
-    auto a = remote_actor();
-    if (!a) {
-      VAST_ERROR("failed to connect:", sys.render(a.error()));
-      return 1;
-    }
-    node = std::move(*a);
-  }
-  self->monitor(node);
-  // Now that we have a node handle, process the commands. If the command
-  // pertains to a local node, we dispatch in this process.
-  if (*cmd == "start") {
-    // Publish the node.
-    auto host =
-      node_endpoint.host.empty() ? nullptr : node_endpoint.host.c_str();
-    auto publish = [&]() -> expected<uint16_t> {
-      if (use_encryption)
-#ifdef VAST_USE_OPENSSL
-        return openssl::publish(node, node_endpoint.port, host);
-#else
-        return make_error(ec::unspecified, "not compiled with OpenSSL support");
+using namespace vast::system;
+
+int main(int argc, char** argv) {
+  // Scaffold
+  configuration cfg{argc, argv};
+  cfg.logger_console = caf::atom("COLORED");
+  caf::actor_system sys{cfg};
+  application app;
+  // Add program commands that run locally.
+  app.add_command<run_start>("start");
```

Comment:
I assume the commands are called `run_*` because of the previous naming. Since they are no longer functions now, but commands instead, we could call them as such: e.g., `start_command`, `stop_command`, etc. Then, we could rename `add_command` to `add`. Here's an example of the API I have in mind:

```cpp
app.add<start_command>("start");
auto import = app.add<import_command>("import");
import->add<reader_command<format::bro::reader>>("bro");
...
```